### DISCUSSION TITLE
Why Don't Samples Compile on Linux

### DISCUSSION THREAD

**Discussion by cuppajoeman** [jrouwe/JoltPhysics#822] [General]
I was reading in the docs and then found this in the samples section: 
![image](https://github.com/jrouwe/JoltPhysics/assets/19198104/e0b8e37e-7a7c-4339-95c1-1ea60bd1af0d)

I am wondering why this is the case, and what would need to be done to have it work natively on linux.

**Comment by jrouwe** [comment]
The sample app uses DirectX, in order to port it to Linux a 2nd Vulkan based renderer would need to get written. This is something I intend to do some day, but for now you can run the samples using Wine as described in the [Build](https://github.com/jrouwe/JoltPhysics/blob/master/Build/README.md) section.

**Reply by cuppajoeman** [reply]
Ok that's great to know, I'd be willing to implement an opengl3 renderer so that it would be supported by older machines as well, then if someone else is interested they would make a Vulkan one  for newer devices.

I won't be at a computer until later tomorrow or the next day. Do you have an interface for the current renderer that could be implemented for the OpenGL renderer?

**Reply by jrouwe** [reply]
In the future I want to make it possible to move part of the simulation to GPU, I don't think this is possible with an OpenGL implementation so I'm only interested in a Vulkan implementation

**Reply by cuppajoeman** [reply]
I think your goals make sense.

Let's say we implemented a Vulkan renderer, now some linux computers can see stuff, other linux computers still wouldn't because their hardware simply doesn't support Vulkan. I think Vulkan is the way to go looking into the future, eventually when everything is running Vulkan this will be the go-to renderer.

Personally my computer's hardware is not there yet, and anyone else who's debugging with Jolt on linux and doesn't have supported Vulkan hardware will need to implement an opengl3 renderer so that they can debug their work. The opengl renderer could act as a fallback renderer for these lower end computers.

 I'm just starting a discussion about this because I will have to implement an opengl3 renderer so that I can debug Jolt stuff myself,  if I need to do that, then someone else will probably have to do it as well, so I was thinking I may as well make a PR if I ended up doing it to save others time. But at the end of the day if you don't think that aligns well with your vision for the repo, it's your call.

Just so we're on the same page, my goal is not to come off as abrasive or anything like that, my goal is to improve the repo as I learn about things coming from my context which is a low spec linux machine :+1: :computer:

**Reply by SirLynix** [reply]
Vulkan on Windows is supported for GPUs that got out in 2014, on Linux support is available for GPUs that got out in 2012 (GCN 1.0, NVidia Kepler, Intel Haswell iGPU).

I really don't think it's worth to implement an outdated API to support GPUs that got out more than ten years ago.  
If your GPU is too old to support Vulkan, you may try [Vulkan implemented on top of OpenGL](https://dragonjoker.github.io/Ashes/).

If we wish to have a renderer that runs anywhere, WebGPU should be considered, [it's way simpler than Vulkan](https://eliemichel.github.io/LearnWebGPU/), the support is coming to web browsers (and [it already has a pretty decent desktop implementation](https://github.com/gfx-rs/wgpu-native) that runs on top of Vulkan/Metal/OpenGL 3.3/GL ES 3.0/WebGL 2.

However WebGPU is not really well-suited for an integration in Jolt since it requires a lib on desktop (like wgpu-native or dawn), I'd suggest to have Vulkan for GPU acceleration in Jolt and WebGPU for demo rendering.

> I don't think this is possible with an OpenGL implementation so I'm only interested in a Vulkan implementation

It's possible but OpenGL is really not suited for this task (it requires a graphics context/swapchain even for compute-only operations), Vulkan does separate rendering and compute pipelines/capabilities.

**Reply by Wunkolo** [reply]
I personally think a Vulkan implementation for both Jolt and its demos is the way to go. It's the most forward-facing cross-platform Graphics API and there are lots of compatibility-layers available for cases where Vulkan cannot be ran natively such as [MoltenVK](https://github.com/KhronosGroup/MoltenVK), [Dozen](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/14766), [Ashes](https://dragonjoker.github.io/Ashes/), and even [Swift Shader](https://github.com/google/swiftshader).

If Jolt's GPU implementation uses Vulkan then the Demos would need that Vulkan support anyways as well. WebGPU is also designed at a level of abstraction where it is not able to access some native Vulkan features that might be interfaced with such as [64-bit integer/float support](https://gpuweb.github.io/gpuweb/wgsl/#integer-types).

**Reply by cuppajoeman** [reply]
Thanks for your input @SirLynix and @Wunkolo. I never knew about the compatibility options for Vulkan, so those are great to know about.

From what I see, if we're were to use Ashes for example we could write the Vulkan renderer with native Vulkan code and then simply switch the include at the top to the ashes header and we would get compatibility for older hardware while also still using Vulkan on machines that allow for that.

@jrouwe What would be your take on using something like ashes while building a Vulkan renderer?

**Reply by SirLynix** [reply]
As a Vulkan confirmed user and big fan, I think it's the right choice for Jolt library as an optional GPU acceleration mode, but not so much for the demos. As it's very explicit and overly complicated for such classical rendering, just compare how many lines are required to draw a triangle [with Vulkan](https://github.com/SaschaWillems/Vulkan/blob/master/examples/triangle/triangle.cpp) and then [with WebGPU](https://github.com/eliemichel/LearnWebGPU-Code/blob/step030/main.cpp). Plus WebGPU runs on more platforms and we may hope to port Jolt Physics demo on web (using emscripten), as I [did myself with my own engine](https://nazara.digitalpulse.software/demo/PhysicsPlayground.html).

> If Jolt's GPU implementation uses Vulkan then the Demos would need that Vulkan support anyways as well

From my understanding, Vulkan would be an implementation detail of Jolt and the demo would not have to use it to use GPU acceleration, it would be completely separate. Of course that means that Jolt Physics demos would not benefit from GPU acceleration when running on the web, which seems completely fine.

> WebGPU is also designed at a level of abstraction where it is not able to access some native Vulkan features that might be interfaced with such as [64-bit integer/float support](https://gpuweb.github.io/gpuweb/wgsl/#integer-types).

This is one of the many reason WebGPU is not suited for GPU acceleration, but it's still perfectly fine as a demo renderer.

> From what I see, if we're were to use Ashes for example we could write the Vulkan renderer with native Vulkan code and then simply switch the include at the top to the ashes header and we would get compatibility for older hardware while also still using Vulkan on machines that allow for that.

You wouldn't even need to change / recompile the code, just [set some environment variables](https://github.com/DragonJoker/Ashes?tab=readme-ov-file#icd-mode) or make the application [load it](https://github.com/DragonJoker/Ashes?tab=readme-ov-file#drop-in-replacement-mode) instead of the Vulkan library on the system.

**Reply by Wunkolo** [reply]
> just compare how many lines are required to draw a triangle [with Vulkan](https://github.com/SaschaWillems/Vulkan/blob/master/examples/triangle/triangle.cpp) and then [with WebGPU](https://github.com/eliemichel/LearnWebGPU-Code/blob/step030/main.cpp)

Something like [Vulkan-Hpp](https://github.com/KhronosGroup/Vulkan-Hpp) would probably be used for both the implementation details in Jolt and the demos to remove much of the trivial fluff of structure-setup, pNext-chains, object-cleanup, and much of the other boilerplate found in the triangle demo. Check out some of the samples there for details. `vk::Unique*` would remove dozens of lines, using `vk::(StructureName)` would remove even more, and some of the other interactions between C++17 features and Vulkan-Hpp can allow even more concise and expressive code.

I also think the existing DX12 demo code lends itself to Vulkan better at the moment as well. It would make an eventual DX12->Vulkan port that much more trivial than a DX12->WebGPU one.

> This is one of the many reason WebGPU is not suited for GPU acceleration, but it's still perfectly fine as a demo renderer.

I can foresee a situation where there is a GPU-Acceleration Jolt-Demo that provides some sort of visualization or interpretation of the underlying Jolt-GPU implementation. During development, these demos would also probably be a testbed of sorts for debugging and diagnosing issues with tools like RenderDoc and Nsight-Graphics which would both require a swapchain workload that eventually touches the Jolt-GPU implementation details. With WebGPU demos, WebGPU would now have to try and interop with Vulkan if there was any in-demo tooling or visualization that accesses this data and would also add this complexity to a gpu-debugging workflow. If the Vulkan implementation uses a feature that WebGPU has no analog for, like 64-bit int/float, then the ability to access or visualize or debug that data from a demo is that much more out of reach.

I also feel like the performance and general evaluation of the Demos is to be _somewhat_ indicative of a shipped title that uses Jolt. WebGPU would be a bit farther away from that than interfacing with DX12/Vulkan directly like most game engines and projects that are evaluating Jolt as their physics-solution would do.

**Reply by jrouwe** [reply]
I agree with @Wunkolo here that it is better to have a Vulkan version first.

Requirements would be that:
* Both DX12 and Vulkan are supported by the demo framework
* I would prefer not to use any 3rd party libraries, although the use of vulkan-hpp doesn't seem that bad inside the Samples folder.
* Inside the Jolt folder only plain DX12/Vulkan code can be used to ensure that users of the library don't get to deal with multiple licenses.

Once I start working on a GPU version an abstraction around 'compute' will need to be written with an implementation for DX12 and Vulkan (and versions for other platforms). This abstraction should live in the Jolt folder. The abstraction should assume that DX12/Vulkan is already fully initialized and should take a minimal set of handles to the needed resources as it will need to integrate into other peoples engines. As the demo framework can render through DX12/Vulkan it can provide those handles and can instantiate the correct abstraction layer based on the current selected renderer (which makes debugging easier as @Wunkolo states).

Given how difficult it is to debug anything created with emscripten at the moment, I'm not sure if I'll ever get around to creating a GPU version for the web.

**Reply by jrouwe** [reply]
FYI: The Jolt samples now run natively on Linux using Vulkan
